packages/pwa-buildpack/lib/WebpackTools/__tests__/MagentoResolver.spec.js

```diff
@@ -1,49 +1,120 @@
+            MagentoResolver.configure({
+                paths: { root: 'fakeRoot' },
+                isEE: true
+            })
+        ).resolves.toEqual({
+            alias: {},
+            modules: ['fakeRoot', 'node_modules'],
+            mainFiles: ['index'],
+            mainFields: ['esnext', 'es2015', 'module', 'browser', 'main'],
+            extensions
+        });
     });
+
+    test('static configure() produces a webpack resolver config with .ce.js extension if backend is not enterprise edition', async () => {
+        const extensions = [
+            '.wasm',
+            '.mjs',
+            '.ce.js',
+            '.js',
+            '.jsx',
+            '.json',
+            '.graphql'
+        ];
+        await expect(
+            MagentoResolver.configure({
+                paths: { root: 'fakeRoot' },
+                isEE: false
+            })
+        ).resolves.toEqual({
+            alias: {},
+            modules: ['fakeRoot', 'node_modules'],
+            mainFiles: ['index'],
+            mainFields: ['esnext', 'es2015', 'module', 'browser', 'main'],
+            extensions
+        });
+    });
+
+    test('static configure() throws if required paths are missing', async () => {
+        await expect(
+            MagentoResolver.configure({ paths: { root: false } })
+        ).rejects.toThrow('paths.root');
+    });
+});
+
+const context = path.resolve(__dirname, './__fixtures__/resolverContext');
+test('creates a promisifed resolver which respects provided config', async () => {
+    const resolver = new MagentoResolver({
+        paths: {
+            root: context
+        }
+    });
+
+    await Promise.all(
+        [
+            // root-relative
+            ['.', 'localModule2.js'],
+            // extension precedence
+            ['localModule1', './localModule1/index.wasm'],
+            // module folder
+            ['depModule1', 'node_modules/depModule1/index.js'],
+            // recursively resolve parent folder deps
+            ['jest', require.resolve('jest')]
+        ].map(([request, expected]) =>
+            expect(resolver.resolve(request)).resolves.toBe(
+                path.resolve(context, expected)
+            )
+        )
+    );
+
+    await expect(resolver.resolve('./missing')).rejects.toThrowError();
 });
 
-test('static configure() produces a webpack resolver config with .ce.js extension if backend is not enterprise edition', async () => {
-    const extensions = [
-        '.wasm',
-        '.mjs',
-        '.ce.js',
-        '.js',
-        '.jsx',
-        '.json',
-        '.graphql'
-    ];
-    await expect(
-        MagentoResolver.configure({ paths: { root: 'fakeRoot' }, isEE: false })
-    ).resolves.toEqual({
-        alias: {},
-        modules: ['fakeRoot', 'node_modules'],
-        mainFiles: ['index'],
-        mainFields: ['esnext', 'es2015', 'module', 'browser', 'main'],
-        extensions
+test('rejects if it has been configured weird', async () => {
```

Comment:
Maybe clarify what constitutes a weird configuration?

Reply:
An unsupported mutation of a private property on the resolver. The resolver is such a key element of asynchronous processing in Webpack that if it doesn't fail loudly and informatively, it's nearly impossible to debug. So I added a use case for unexpected state within the resolver.